README.rst
====
Leela-Dashboard
====
A simple dashboard for Leela (http://github.com/locaweb/leela-client)

Configuration
=============
All configuration for the Leela Dashboard is in the yaml file on /etc/leela/leeladashboard.yml. There it is the metrics section where you can define which leela metrics you want to display for each cluster/server and if necessary you can define Cluster Groups so you can aggregate the metrics with all servers defined on that cluster group.


Example:

metrics:
    - loadavg.1
    - memory.main.used
    - cpu.cpu.user
    - apache_default.req/s

cluster_group:
    - ApacheClusters

ApacheClusters:
    - cluster01
    - cluster02

cluster01:
    - myserver01
    - myserver02

cluster02:
    - myserver03
    - myserver04
    - myserver08


Version
======
$version 0.2

License
======
Apache 2.0


debian/changelog
leela-dashboard (0.2) unstable; urgency=low

  * fix static files in *.html
  * fix metric var in cluster.html
  * new version

 -- Rogerio Goncalves <rogerio.goncalves@locaweb.com.br>  Fri, 13 Jun 2014 14:42:45 +0000

leela-dashboard (0.1) unstable; urgency=low

  * Initial Release.

 -- Andre Ferraz <andre.ferraz@locaweb.com.br>  Mon, 25 Mar 2013 14:42:45 +0000


src/app/leeladash.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from flask import Flask
from flask import render_template
from flask import url_for
from flask import request
from flask import redirect

import yaml

app = Flask(__name__)

cvalues = {}
with open('/etc/leela/leeladashboard.yml', 'r') as conf:
    cvalues = yaml.load(conf)

@app.route('/')
def index():
    return render_template('index.html',
            cvalues = cvalues)

@app.route('/static/<path:filename>')
def base_static(filename):
        return send_from_directory(app.root_path + '/static/', filename)

@app.route('/server')
def server():
    servername = request.args.get('servername', None)
    week       = request.args.get('week', False)
    return render_template('server.html',
            cvalues    = cvalues,
            servername = servername)

@app.route('/cluster_group/<cluster>', methods=['POST', 'GET'])
@app.route('/cluster_group/<cluster>/<farm>', methods=['POST', 'GET'])
def pulley(cluster=None, farm=None):
    legend = request.args.get('legend', False)
    week   = request.args.get('week', False)
    if cluster is None:
        return redirect(url_for('index'))
    else:
        if farm is None:
            return render_template('cluster_overview.html', 
                    cluster = cluster,
                    farm    = farm,
                    legend  = legend,
                    cvalues = cvalues,
                    week    = week)
        else:
            return render_template('cluster.html',
                    cluster = cluster,
                    legend  = legend,
                    cvalues = cvalues,
                    week    = week,
                    farm    = farm)

if __name__ == '__main__':
    app.run(host='0.0.0.0', debug=True, port=8080)


src/app/static/home.js
(function($) {
  $(function() {
    var controls = $(".banner-controls").first();
    controls.parents(".box").first().children("h2").first().append(controls);

    $("#view-all-banners-link").click(function(e) {
      var alternativeText = $(this).data("alternative-text");
      $(this).data("alternative-text", $(this).html());
      $(this).html(alternativeText);

      $(".banners-slider").toggle();
      $(".banner-navigation").toggle();
      var container = $(".all-banners-container");

      if ($(".banners-slider").is(":visible") == false) {
        container.load($(this).attr("href"), function() {
          container.removeClass("loading");
          container.find(".banner-actions").hide();
        });
        container.show();
        container.addClass("loading");
      } else {
        container.hide();
      }

      e.preventDefault();
      return false;
    });
  });
/*
  $.fn.bannerGallery = function() {
    var index = 0;
    var pages = 0;
    var width = 959;
    var auto = null;

    var next = function(e) {
      return function() {
        clearInterval(auto);
        if (index < pages) {
          index++;
          var _left = index * width;

          if (index < pages)
            e.find('ul').animate({marginLeft: -_left}, {queue: false, duration: 500});
          else
            index--;
        }
      };
    };

    var prev = function(e) {
      return function() {
        clearInterval(auto);
        if (index > 0) {
          index--;
          var _left = index * width;
          $(e).children('ul').animate({marginLeft: -_left}, {queue: false, duration: 500});
        }
      };
    };

      return this.each(function() {
      $(this).find('.link-next').click(next($(this)));
      $(this).find('.link-prev').click(prev($(this)));

      index = 0;
      pages = $(this).find('li').length;
      width = $(this).find('li').first().innerWidth();
      $(this).find('li').height($(this).innerHeight());
        });
    }*/

  $.fn.bannerGallery = function(_options){
    var defaults = {
      gallery: null,
      index:0,
      pages:0,
      width:0,
      isLeft: false,
      isExec: true,
      autoInterval:null,
      auto: function(){
        clearInterval(options.autoInterval);
        if(options.isExec) {
          options.autoInterval = setInterval(function(){
            if(options.isLeft) {
              options.gallery.find('.link-prev').trigger('click');
            } else {
              options.gallery.find('.link-next').trigger('click');
            }

          }, 10000);
        }
      },

      left: function(){
        clearInterval(options.autoInterval);
        
        if (options.index > 0) {
          options.index--;
          var _left = options.index * options.width;
          options.gallery.find('ul').animate({marginLeft: -_left}, {queue: false, duration: 500, easing:'easeOutQuint', complete: function(){
            options.auto();
          }});

          if(options.index === 0)
            options.isLeft = false;
        }
      },

      right: function(){
        clearInterval(options.autoInterval);

        if (options.index < options.pages) {
          options.index++;
          var _left = options.index * options.width;
          if (options.index < options.pages) {
            options.gallery.find('ul').animate({marginLeft: -_left}, {queue: false, duration: 500, easing:'easeOutQuint', complete: function(){
              options.auto();
            }});
          } else
            options.index--;
        }

        if(options.index === (options.pages-1)) {
          options.isLeft = true;
        }
      },

      mouseenter: function(){
        clearInterval(options.autoInterval);
        options.isExec = false;
      },

      mouseleave: function(){
        options.isExec = true;
        options.auto();
      }
    };

    var options = $.extend(defaults, _options);

    return this.each(function(i, e){
      $(e).find('.link-next').bind('click', options.right);
      $(e).find('.link-prev').bind('click', options.left);

      options.pages = $(e).find('ul li').length;
      options.width = $(e).find('ul li:eq(0)').innerWidth();
      $(e).find('ul li').height($(e).innerHeight());
      $(e).bind('mouseenter', options.mouseenter).bind('mouseleave', options.mouseleave);
      options.gallery = $(e);

      options.auto();
    });

  };
})(jQuery);



src/app/static/leela/flotr2_backend.js
/*jslint continue: true, vars: true, indent: 2*/

// All Rights Reserved.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

var LEELA;

if (LEELA === undefined) {
  LEELA = {};
}

if (LEELA.backend === undefined) {
  LEELA.backend = {};
}

LEELA.backend.flotr2 = function (root) {
  "use strict";

  var zoom;
  var curdata;
  var curopts;
  var r_thread;
  var fsttime = true;

  var build_options = function (options) {
    var myopts = { HtmlText : false, // allows labelsAngle
                   xaxis: { title: LEELA.f.getprop(options, ["xaxis", "title"]),
                            tickFormatter: LEELA.f.getprop(options, ["xaxis", "labels", "formatter"], Flotr.defaultTickFormatter),
                            mode: "time",
                            timeFormat: "%d %b, %H:%M",
                            timeMode: "local",
                            timeUnit: "second",
                            max: LEELA.f.getprop(options, ["xaxis", "max"]),
                            min: LEELA.f.getprop(options, ["xaxis", "min"]),
                            noTicks: options.width / 42,
                            labelsAngle: 45
                          },
                   yaxis: { autoscale: true,
                            title: LEELA.f.getprop(options, ["yaxis", "title"]),
                            tickFormatter: LEELA.f.getprop(options, ["yaxis", "labels", "formatter"], Flotr.defaultTickFormatter),
                            max: LEELA.f.getprop(options, ["yaxis", "max"]),
                            min: LEELA.f.getprop(options, ["yaxis", "min"])
                          },
                   title: options.title,
                   subtitle: options.subtitle,
                   selection: { mode: "x",
                                fps: 30
                              },
                   legend: { position: "nw",
                             //container: LEELA.f.getprop(options, ["legend", "container"], null)
                           }
                 };
    if (zoom !== undefined) {
      myopts.xaxis.min = zoom[0];
      myopts.yaxis.min = zoom[1];
      myopts.xaxis.max = zoom[2];
      myopts.yaxis.max = zoom[3];
    } else {
      if (myopts.xaxis.max === undefined) {
        delete (myopts.xaxis.max);
      }
      if (myopts.xaxis.min === undefined) {
        delete (myopts.xaxis.min);
      }
      if (myopts.yaxis.min === undefined) {
        delete (myopts.yaxis.min);
      }
      if (myopts.yaxis.max === undefined) {
        delete (myopts.yaxis.max);
      }
    }
    return (myopts);
  };

  var do_redraw = function () {
    if (curdata !== undefined && curopts !== undefined) {
      Flotr.draw(root, curdata, build_options(curopts));
    }
  };

  var redraw = function () {
    if (r_thread !== undefined) {
      clearTimeout(r_thread);
      r_thread = undefined;
    }
    r_thread = setTimeout(function () {
      Flotr.draw(root, curdata, build_options(curopts));
      r_thread = undefined;
    }, (fsttime ? 100 : 300));
  };

  var zoomin = function (x0, y0, x1, y1) {
    zoom = [x0, y0, x1, y1];
    redraw();
  };

  var zoomout = function () {
    if (zoom !== undefined) {
      zoom = undefined;
      redraw();
    }
  };

  Flotr.EventAdapter.observe(root, "flotr:select", function (area) {
    if (Math.abs(area.x2 - area.x1) > 60) { // 60: assuming a time series
      zoomin(area.x1, area.y1, area.x2, area.y2);
    }
  });

  Flotr.EventAdapter.observe(root, "flotr:click", function () {
    zoomout();
  });

  var cspline_i = function (x, xk_1, yk_1, xk, yk, xk1, yk1, xk2, yk2) {
    // http://en.wikipedia.org/wiki/Cubic_Hermite_spline#Finite_difference
    var m0  = (yk1 - yk) / (2 * (xk1 - xk)) + (yk - yk_1) / (2 * (xk - xk_1));
    var m1  = (yk2 - yk1) / (2 * (xk2 - xk1)) + (yk1 - yk) / (2 * (xk1 - xk));
    // http://en.wikipedia.org/wiki/Cubic_Hermite_spline#Interpolation_on_.28xk.2C_xk.2B1.29
    var t   = (x - xk) / (xk1 - xk);
    var t2  = t * t;
    var t3  = t2 * t;
    var h00 = 2 * t3 - 3 * t2 + 1;
    var h10 = t3 - 2 * t2 + t;
    var h01 = 3 * t2 - (2 * t3);
    var h11 = t3 - t2;
    var y   = h00 * yk + h10 * (xk1 - xk) * m0 + h01 * yk1 + h11 * (xk1 - xk) * m1;
    return ([x, y]);
  };

  var chspline = function (data) {
    // http://en.wikipedia.org/wiki/Cubic_Hermite_spline
    var ndata = [];
    var len   = data.length;
    var res   = 25;

    var k;
    var u;
    var xy;
    for (k = 0; k < (len - 1); k += 1) {
      var x    = data[k][0];
      var xk_1 = (data[k - 1] || [0])[0];
      var yk_1 = (data[k - 1] || [0, 0])[1];
      var xk   = data[k][0];
      var yk   = data[k][1];
      var xk1  = data[k + 1][0];
      var yk1  = data[k + 1][1];
      var xk2  = (data[k + 2] || [0])[0];
      var yk2  = (data[k + 2] || [0, 0])[1];
      var s    = (xk1 - xk) / res;
      for (u = 0; u < res; u += 1) {
        x += s;
        xy = cspline_i(x, xk_1, yk_1, xk, yk, xk1, yk1, xk2, yk2);
        ndata.push([xy[0], xy[1]]);
      }
    }

    return (ndata);
  };

  var format_s = function (json, fmap, options) {
    var series   = [];
    var deffmt   = function (k) {
                     var max  = LEELA.f.max(LEELA.f.map(json.results[k].series, LEELA.f.snd));
                     var min  = LEELA.f.min(LEELA.f.map(json.results[k].series, LEELA.f.snd));
                     var mean = LEELA.f.mean(LEELA.f.map(json.results[k].series, LEELA.f.snd));
                     var avg  = LEELA.f.avg(LEELA.f.map(json.results[k].series, LEELA.f.snd));
                     var tmp  = "[";
                     tmp += "max:  "+ max.toFixed(1) + ", ";
                     tmp += "min:  "+ min.toFixed(1) + ", ";
                     tmp += "mean: "+ mean.toFixed(1) + ", ";
                     tmp += "avg:  "+ avg.toFixed(1) + "]";
                     return(k + " " + tmp);
                   };
    var labelfmt = LEELA.f.getprop(options, ["chart", "labels", "formatter"], deffmt);
    var k;
    for (k in json.results) {
      if (json.results.hasOwnProperty(k)) {
        series.push({ label: json.results[k].label || labelfmt(k),
                      data: fmap(json.results[k].series)
                    });
      }
    }
    return (series);
  };

  var render = function (json, options) {
    var type = LEELA.f.getprop(options, ["chart", "type"], "spline");
    curdata  = (type === "spline" ? format_s(json, chspline, options) : format_s(json, LEELA.f.id, options));
    curopts  = options;
    zoom     = undefined;
    redraw();
  };

  return ({"render": render,
           "zoomin": zoomin,
           "zoomout": zoomout
          });
};
